### DISCUSSION TITLE
Autocomplete compile comands based on compile-history

### DISCUSSION THREAD

**Discussion by SpyrosRoum** [minad/vertico#531] [Q&A] [answered]
Hello, first of all I'm not sure if vertico is the correct place to ask, but if anyone I'm sure you would know how to direct me.

What I want to do is to get auto-complete in the compile-command prompt that will auto-suggest entries from `compile-history`. My searches have been failing me, but I believe this has to be possible somehow. 
`compile-history` already has a list with values, pressing the up/down arrows even navigates this list by default, I'm just not sure how to hook it up to the vertico ui.

Any help would be greatly appreciated, thank you in advance

**Comment by SpyrosRoum** [comment, marked as answer]
To my absolute surprise I was able to do this myself after reading through vertico's readme and the linked emacs docs for completions.. I'll leave my solution here since I don't want this to be another "nvm I fixed it" post, and I'll leave the issue open in case you know of any better solutions, but feel free to just close.

```elisp
;; compilation-read-command uses `read-shell-command` by default, which doesn't use
;; completion at all. So I overwrite it to use `completing-read` instead, which seems to work great.
(defun compilation-read-command (command)
  (completing-read "Compile command: " compile-history
    nil nil command
    (if (equal (car compile-history) command)
      '(compile-history . 1)
      'compile-history)))
```

**Reply by minad** [reply]
I converted to a Q&A discussion for others to see the solution.

**Comment by rudolf-adamkovic** [comment]
@SpyrosRoum 

This is fun!  Thank you for sharing.  

Here is a version that shows all candidates, like say `M-x`:

```elisp
(defun compilation-read-command (command)
  (completing-read "Compile command: " compile-history
                   nil nil nil
                   (if (equal (car compile-history) command)
                       '(compile-history . 1)
                     'compile-history)
                   command))
```

**Comment by rudolf-adamkovic** [comment]
**Note.** It would perhaps be a good idea to contribute a patch to Emacs where the user could set the `compilation-read-command` to a function like the ones above (to avoid risky overrides).

**Reply by SpyrosRoum** [reply]
> **Note.** It would perhaps be a good idea to contribute a patch to Emacs where the user could set the `compilation-read-command` to a function like the ones above (to avoid risky overrides).

Hey, I wouldn't mind doing that but I'm relatively new to elisp so I'm not sure how that would look like. Do you have of any other mechanisms that work like that currently that I can check out?

**Reply by rudolf-adamkovic** [reply]
I think the best next step would be to bring this (IMO, great) idea up for discussion over on the `emacs-devel` mailing list.

**Comment by rudolf-adamkovic** [comment]
And here is an `advice`, so that _Help_ shows

> This function has :override advice: ‘my-compilation-read-command’.

```elisp
(defun my-compilation-read-command (command)
  (completing-read "Compile command: " compile-history
                   nil nil nil
                   (if (equal (car compile-history) command)
                       '(compile-history . 1)
                     'compile-history)
                   command))

(advice-add #'compilation-read-command
            :override
            #'my-compilation-read-command)
```
